### PR TITLE
fix(service-worker): ngsw.json navigationUrls breaks all service worker versions < 6.0.0

### PR DESCRIPTION
**Pull request by Serginho**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [X] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: 27209
https://github.com/angular/angular/issues/27209

## What is the new behavior?

Fix the bug, I explain how I fixed in the issue comments.

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

**Comment by gkalpak** [comment]
Thx, @Serginho! I am still not convinced this is the right approach, though.
Let's discuss in https://github.com/angular/angular/issues/27209#issuecomment-441212262.

**Comment by Serginho** [comment]
@gkalpak I updated the pull request with new changes. WDYT?

**Review by gkalpak — CHANGES_REQUESTED** [review]
Ooops, seems like I had added review comments, but forgot to push the "Submit review" button. Thx for the ping.

File: packages/service-worker/worker/src/app-version.ts [hunk 1]
```diff
@@ -15,6 +15,13 @@ import {IdleScheduler} from './idle';
 import {Manifest} from './manifest';
 
 
+const DEFAULT_NAVIGATION_URLS = [
```
[the remark this answers is not in the document]
Comment (Serginho) [reply packages/service-worker/worker/src/app-version.ts]: Change it

File: packages/service-worker/worker/src/app-version.ts [hunk 2]
```diff
@@ -85,6 +92,11 @@ export class AppVersion implements UpdateSource {
                                   this.scope, this.adapter, config, this.database,
                                   `ngsw:${config.version}:data`));
 
+
+    // This keeps backwards compatibility with app versions without navigation urls.
+    // Fix: https://github.com/angular/angular/issues/27209
```
[the remark this answers is not in the document]
Comment (Serginho) [reply packages/service-worker/worker/src/app-version.ts]: Thx 😃

**Review by gkalpak — CHANGES_REQUESTED** [review]
Thx for following up, @Serginho 👍 The code LGTM.
I just realized that it would be good to have some tests for the changes. Would you be interested in looking into that?

**Comment by mhevery** [comment]
Can we get the tests green.

**Comment by gkalpak** [comment]
Rebasing on master should take care of the broken tests. But the main problem is that we don't have tests for the fix (see https://github.com/angular/angular/pull/27244#pullrequestreview-182866614).

@Serginho, are you still interested in moving this forward?

**Comment by Serginho** [comment]
@gkalpak Yes, I'm interested in. I've had lot of work last weeks.

**Comment by gkalpak** [comment]
Awesome, @Serginho  :tada:
Let me know if need help/guidance with the tests.

File: packages/service-worker/worker/test/happy_spec.ts [hunk 3]
```diff
@@ -66,6 +65,64 @@ const brokenManifest: Manifest = {
...
+  index: string;
+  assetGroups?: AssetGroupConfig[];
+  dataGroups?: DataGroupConfig[];
+  hashTable: {[url: string]: string};
+}
+
+// To simulate versions < 6.0.0
+const manifestOld: ManifestV5 = {
+  configVersion: 1,
+  appData: {
```
Comment (gkalpak) [inline packages/service-worker/worker/test/happy_spec.ts]: `appData` should not be needed.

File: packages/service-worker/worker/test/happy_spec.ts [hunk 4]
```diff
@@ -66,6 +65,64 @@ const brokenManifest: Manifest = {
...
+    version: 'original',
+  },
+  index: '/foo.txt',
+  assetGroups: [
+    {
```
Comment (gkalpak) [inline packages/service-worker/worker/test/happy_spec.ts]: Are asset groups necessary to reproduce the bug? We should keep the manifest as minimal as possible.

File: packages/service-worker/worker/test/happy_spec.ts [hunk 5]
```diff
@@ -192,7 +249,8 @@ const manifestHash = sha1(JSON.stringify(manifest));
 const manifestUpdateHash = sha1(JSON.stringify(manifestUpdate));
 
 (function() {
-  // Skip environments that don't support the minimum APIs needed to run the SW tests.
+  // Skip environments that don't support the minimum APIs needed to run the SW
+  // tests.
```
Comment (gkalpak) [inline packages/service-worker/worker/test/happy_spec.ts]: This change (and similar ones below) seems irrelevant 😕 
Please, revert.

File: packages/service-worker/worker/test/happy_spec.ts [hunk 6]
```diff
@@ -989,6 +1049,36 @@ const manifestUpdateHash = sha1(JSON.stringify(manifestUpdate));
         expect(await requestFoo('only-if-cached', 'same-origin')).toBe('this is foo');
         expect(await requestFoo('only-if-cached', 'no-cors')).toBeNull();
       });
+      // Test this bug: https://github.com/angular/angular/issues/27209
```
Comment (gkalpak) [inline packages/service-worker/worker/test/happy_spec.ts]: Insert empty line between testcases.

File: packages/service-worker/worker/test/happy_spec.ts [hunk 7]
```diff
@@ -989,6 +1049,36 @@ const manifestUpdateHash = sha1(JSON.stringify(manifestUpdate));
...
+          async() => {
+            const cache = new CacheDatabase(scope, scope);
+            const table = await cache.open('control');
+
+            const hash = hashManifest(<Manifest>manifestOld);
+
+            // Initialize cache with a state like angular 5.x
+            await table.write('manifests', {[hash]: manifestOld});
+            await table.write('assignments', {'default': hash});
+            await table.write('latest', {latest: hash});
```
Comment (gkalpak) [inline packages/service-worker/worker/test/happy_spec.ts]: Instead of manually replicating these operations, couldn't you use a test server-state (as in [here][1])?

[1]: https://github.com/angular/angular/blob/ec1bc0c770147f70c4024a976b34f0e8c6f068a0/packages/service-worker/worker/test/happy_spec.ts#L227-L232

File: packages/service-worker/worker/test/happy_spec.ts [hunk 8]
```diff
@@ -989,6 +1049,36 @@ const manifestUpdateHash = sha1(JSON.stringify(manifestUpdate));
...
+            expect(await makeRequest(scope, '/foo.txt')).toEqual('this is foo');
+            await driver.initialized;
+            scope.updateServerState(serverUpdate);
+            expect(await driver.checkForUpdate()).toEqual(true);
+
+            const cachedManifests = await table.read('manifests') as{[hash: string]: Manifest};
+            Object.keys(cachedManifests).forEach(hash => {
+              expect(cachedManifests[hash].navigationUrls).toBeTruthy();
+            });
+            expect(cachedManifests[hash].navigationUrls).toEqual([
```
Comment (gkalpak) [inline packages/service-worker/worker/test/happy_spec.ts]: These assertions are not important. The important thing is that the SW does not break and can still serve the files (and update to newer versions) as expected.

**Review by gkalpak — CHANGES_REQUESTED** [review]

File: packages/service-worker/worker/test/happy_spec.ts (hunk 8 above)
Comment (Serginho) [reply packages/service-worker/worker/test/happy_spec.ts]: Ok, but I think this one is interesting cause check the bug:
```
Object.keys(cachedManifests).forEach(hash => {
              expect(cachedManifests[hash].navigationUrls).toBeTruthy();
});
```

Comment (gkalpak) [reply packages/service-worker/worker/test/happy_spec.ts]: The bug is not that `navigationUrls` was falsy. That is an implementation detail. The bug was that the SW crashed.
If we can verify that the SW does not crash and can update correctly, then it doesn't matter whether `navigationUrls` is truthy or not, right?

**Comment by Serginho** [comment]
@gkalpak All review changes are done. I think this is the solutions you are looking for. I hope the PR will be merged this time.

**Review by gkalpak — APPROVED** [review]
Awesome, @Serginho!
Have you verified that the test does indeed fail without the fix? Otherwise LGTM 👍

**Comment by gkalpak** [comment]
Can you also rebase on master to get CI green?

**Comment by Serginho** [comment]
> Awesome, @Serginho!
> Have you verified that the test does indeed fail without the fix? Otherwise LGTM 👍

yes, if you delete `manifest.navigationUrls = manifest.navigationUrls || BACKWARDS_COMPATIBILITY_NAVIGATION_URLS;` the test fails.

**Comment by Serginho** [comment]
I think looks good now

**Comment by Serginho** [comment]
@gkalpak It's ready to merge! 🎉

**Review by gkalpak — APPROVED** [review]

**Comment by gkalpak** [comment]
Not so fast 😛
Can you please squash the commits (or make the second one a fixup commit). E.g.:

```
git checkout service_worker_navigationurls_fix
git reset HEAD~1

# Then squash:
git commit --all --amend --no-edit

# Or make fixup commit:
git commit --fixup HEAD
```

**Comment by Serginho** [comment]
@gkalpak I prefer  `git reset $(git merge-base master service_worker_navigationurls_fix)`

Done.

**Comment by gkalpak** [comment]
Not sure how `git merge-base master service_worker_navigationurls_fix` is relevant for the task at hand, but whatever works for you :grin:
Thx for rebasing :heart:

**Comment by gkalpak** [comment]
Thx for staying with me, @Serginho 👍 I've marked this for merge :tada:

(I know the process may feel more complicated than necessary and even frustrating at times. Thx for pushing through 🏋️)

**Comment by Serginho** [comment]
@gkalpak 

> Not sure how `git merge-base master service_worker_navigationurls_fix` is relevant for the task at hand, but whatever works for you 😁
> Thx for rebasing ❤️
It's not relevant, I use this cause I have some automated scripts.

> Thx for staying with me, @Serginho 👍 I've marked this for merge 🎉
> 
> (I know the process may feel more complicated than necessary and even frustrating at times. Thx for pushing through 🏋️)
I'm so exited about contributing to this framework. I think I'll do it again in the future, cause I develop new features in a private fork of this package.
